src/mender-update/context/context.cpp

```diff
@@ -243,30 +242,24 @@ static error::Error FilterProvides(
 		string escaped;
 		// Potential to escape every character, though unlikely.
 		escaped.reserve(to_clear.size() * 2);
-		// Notable exception: '*', since it has special handling as a glob character.
-		string meta_characters {".^$+()[]{}|?"};
+
+		bool wildcard = false;
 		for (const auto chr : to_clear) {
 			if (chr == '*') {
-				// Turn every '*' glob wildcard into '.*' regex wildcard.
-				escaped.push_back('.');
-			} else if (any_of(meta_characters.begin(), meta_characters.end(), [chr](char c) {
-						   return chr == c;
-					   })) {
-				// Escape every regex special character except '*'.
-				escaped.push_back('\\');
+				// Acts as a regex wildcard, break so `*` isn't included,
+				// this allows us to use the substring to match like a regex
+				wildcard = true;
+				break;
 			}
 			escaped.push_back(chr);
 		}
 
-		regex compiled;
-		auto err = error::ExceptionToErrorOrAbort(
-			[&compiled, &escaped]() { compiled.assign(escaped, regex_constants::basic); });
-		// Should not be possible, since the whole regex is escaped.
-		AssertOrReturnError(err == error::NoError);
-
 		set<string> keys;
 		for (auto provide : to_modify) {
-			if (regex_match(provide.first.begin(), provide.first.end(), compiled)) {
+			if (escaped == provide.first) {
+				keys.insert(provide.first);
+				// Match escaped 'wildcard substring'
+			} else if (provide.first.rfind(escaped, 0) == 0 && wildcard) {
```

Comment:
I think this could give false negatives. Imagine the string `test_test_key`. If the search string has `test_*`, then it will not match because `rfind` finds the second occurrence first, which is not at the zero index.

A quick fix is to use `find` instead. On the other hand, `find` is overkill in this case, because we know that the string will always be at the start, so perhaps it's better to use `provide.first.compare(0, escaped.size(), escaped) == 0` (see [`compare`](https://en.cppreference.com/w/cpp/string/basic_string/compare)).

Reply:
This won't happen, because I specified `pos` (the second argument) as 0. This sets the position of the last character to be considered the beginning of a match to index 0, but this just seems like find with extra steps, so `compare` seems better and I will use that instead.